---
title: "Genetic Linkage Map - *Mimulus*"
author: "Gabriel Mamédio, Igor Araújo, Jenifer Camila Godoy e Talieisse Fagundes"
date: "6/7/2021"
output: html_document
---


</p>

### 1. Description of Data (Published Map)


<div align="justify"> Fishman *et al*. (2001) constructed and analyzed a linkage map based on an F2 hybrid population composed of 526 individuals (N = 526). First, the F2 hybrid population was generated by crossing  *Mimulus guttatus* and *Mimulus nasutus* inbred lines. The *M. guttatus* parental line was derived from an annual highly outcrossing population and was formed by more than five generations of selfing with single seed descent) (IM62). The *M. nasutus* parental line was derived from a wild population and maintained for several generations in the greenhouse through autonomous self-fertilization (SF5.4). Later, this F2 population was genotyped for 255 molecular markers (27 microsatellites; 4 gene-based markers; and 224 AFLPs). </div>

</p>

<div align="justify"> The genetic linkage map was constructed using the software MAPMAKER 3.0 (Lander *et al*., 1987). They used the GROUP command with the Kosambi mapping to organize markers into linkage groups (two-point linkage criteria: minimum LOD 6.0 and maximum distance between markers of 37 cM) and the ORDER command to map the most likely order for each group. The multipoint ordering used a threshold of LOD 3.0 to find a starting subset of five markers and to place markers in a first-round and then tried to place the remaining markers with a LOD threshold of 2.0. The unreliable markers, along with any linked but unplaced markers, were then individually evaluated using the TRY, COMPARE, MAP, and RIPPLE commands to generate and compare alternative orders. The final map consists of 174 markers distributed on 14 linkage groups with a total length of 1780 cM (Kosambi). The 14 framework linkage groups correspond to the haploid chromosome number of these *Mimulus* species.  </div>

</p>

<div align="justify"> The linkage map presented by them provides a framework for studying the genetic architecture of an evolutionarily important adaptation and also allows the direct examination of interactions between genomes in the early stages of speciation. In addition, the map reveals localized and directional segregation distortion, which has important implications for speciation and hybridization. </div>


</p>

### 2. Description of Data (Map Update)

<div align="justify"> After the publication of Fishman *et al*. (2001), more loci were genotyped. Now, we will build a new map using the entire dataset using the software OneMap (Margarido *et al*., 2007).  OneMap was the first software developed to simultaneously estimate the linkage and linkage phases (integrated genetic map) between markers in outcrossing species (F1). It has two tutorials that teach step by step how to construct a genetic linkage map in inbreeding and outcrossing populations. You can consult the tutorials here (https://github.com/augusto-garcia/onemap). 


### 2.1 Importing Data

<div align="justify">  For F2s, two input formats are accepted: OneMap file format or MAPMAKER file format. Both types of raw file can contain phenotypic information, but this will not be used during map construction, that requires only genotypic information (made available by molecular markers). As the dataset file is in MAPMAKER format, we used the **read_mapmaker** function to import it to OneMap. </div>

</p>

```{r setseed, echo=FALSE, message=FALSE, warning=FALSE}
set.seed(3333)
```

```{r packages, echo=TRUE, message=FALSE, warning=FALSE}
library("onemap")
library("gridExtra")
```

```{r data1, echo=TRUE, message=FALSE, warning=FALSE}
data_map = read_mapmaker(file="m_feb06.raw")
```

```{r data2, echo=TRUE, message=FALSE, warning=FALSE}
data_map
```

<div align="justify"> Note that the data consists of a sample of 287 individuals genotyped for 418 markers with 11% of missing data. In relation to the total number of marks,  213 markers are codominant (AA, AB, or AB), 92 markers are dominant in one parent (not AA or AA), and 113 markers are dominant in the other parent (not BB or BB). In addition, there is no chromosome and physical position information and there is phenotypic information for 16 traits in the data set, that can be used for QTL mapping.  </div>

</p>

### 2.2 Visualization of Raw Data

<div align="justify"> Using the OneMap it is possible to see the graphic of raw data and the number of loci within each type of marker. These graphics show that the proportion of codominant and dominant markers is similar.  </div>


</p>

```{r plot1, echo=TRUE, fig.align='center', fig.height=5, fig.width=8, warning=FALSE}
plot(data_map)
```

```{r plot2, echo=TRUE, fig.align='center', fig.height=5, fig.width=8, warning=FALSE}
plot_by_segreg_type(data_map)
```

</p>

### 2.3 Filtering of Markers

</p>

#### 2.3.1 Redundant Markers:

<div align="justify"> Before the map construction, two steps of filtration were performed. In the first filtering, the markers with redundant information were removed using the **find_bins** and **create_data_bins** functions. Redundant markers are markers that have the same genotypic information, usually because recombination events did not happen between them. In this case, these markers are not informative and increase the computational effort during the map building. Therefore, they can be removed for map building. Once the maps are finished, they can be added again. </div>

</p>

```{r red1, echo=TRUE, message=FALSE, warning=FALSE}
bins = find_bins(data_map, exact = FALSE)
bins
```

```{r red2, echo=TRUE, message=FALSE, warning=FALSE}
data_bins = create_data_bins(data_map, bins)
data_bins
```

<div align="justify">  In this step, two redundant markers were removed and now the dataset has 416 markers. </div>

</p>

#### 2.3.2 Segregation Test:

<div align="justify"> In the second filtering, the markers with Mendelian segregation deviations were identified using the **test_segregation** function. This function performs the chi-square test using global alpha ($\alpha$) of Bonferroni correction for multiple tests. With the **select_segreg function**, markers without segregation deviations were selected. Using the **plot** function from the OneMap package, it is possible to see how many markers are distorted under Bonferroni’s criterion. In this case, 14% of markers have segregation deviation. </div>

</p>

<div align="justify"> It is not recommended to use these markers to build the genetic linkage maps, because generally the statistical models for the map building process follow Mendel’s law as an assumption. However, they can be used to saturate the map.  </div>

</p>

```{r seg1, echo=TRUE, message=FALSE, warning=FALSE}
data_seg = test_segregation(data_bins)
a = print(data_seg) 
head(a, 10)
```

```{r seg2, echo=TRUE, message=FALSE, warning=FALSE}
Bonferroni_alpha(data_seg)
```

```{r seg4, echo=TRUE, message=FALSE, warning=FALSE}
select_segreg(data_seg, distorted = FALSE)
```

```{r seg6, echo=TRUE, message=FALSE, warning=FALSE}
no_dist = select_segreg(data_seg, distorted = FALSE, numbers = TRUE)
no_dist
```

```{r seg3, echo=TRUE, fig.align='center', fig.height=5, fig.width=8, warning=FALSE}
plot(data_seg)
```

<div align="justify"> In this step, 60 markers with segregation deviation were removed. Now the dataset has 356 markers that will be used in the next analyses.  </div>

</p>

### 3. Construction of the Genetic Linkage Maps

<div align="justify"> The main steps to construct linkage maps are (i) linkage analysis of markers,  (ii) assigning markers to linkage groups, (iii) ordering of the markers within linkage groups, and (iv) estimation of genetic distances. </div>

</p>

### 3.1 Linkage Analysis:

<div align="justify"> Linkage analysis of molecular markers is based on the recombination fraction and odds ratios expressed in logarithm (logarithm of odds - LOD). The recombination fraction is calculated between all pairs of markers (two-point tests) and usually values smaller than 0.5 are used to declare the linkage. The LOD score is the statistic used to evaluate the significance of the test and its value takes into consideration the number of tests performed, which depends of the number of markers. For example, a LOD score of 3 between two markers, indicates that linkage is 1000 times more likely than no linkage. </div>

</p>

<div align="justify"> The recombination fractions between all pairs of markers (two-point tests) were estimated using  **rf_2pts** function. Here, we consider the maximum recombination fraction equal to 0.5 and the LOD equal to 5.42 as suggested by the **suggest_lod** function. </div>

</p>

```{r rf1, echo=TRUE, message=FALSE, warning=FALSE}
LOD_sug = suggest_lod(data_bins)
LOD_sug
```

```{r rf2, echo=TRUE, message=FALSE, warning=FALSE}
twopts = rf_2pts(input.obj = data_bins, LOD = LOD_sug, max.rf = 0.5)
```

</p>

### 3.2 Assigning Markers to Linkage Groups:

<div align="justify"> Based on the values of recombination fraction and LOD, the markers were assigned to linkage groups using **make_seq**  and **group** functions.  Linkage groups represent chromosomal segments or entire chromosomes. </div> 

</p>

<div align="justify"> In this step, we used the vector of markers without segregation distortion. Note that 14 linkage groups were formed. </div> 


</p>

```{r rf3, echo=TRUE, message=FALSE, warning=FALSE}
mark_no_dist = make_seq(twopts, c(no_dist))
lgs = group(mark_no_dist)
```


</p>


### 3.3 Ordering of Linkage Groups and Estimation of Genetic Distances:

</p>

#### 3.3.1 Estimation of Genetic Distances:

<div align="justify">  The final step of the construction of a genetic linkage map involves the estimation of the markers’ order and distance within linkage groups. As the recombination fraction is not additive, mapping functions are required to estimate the distance of markers. Several functions have been proposed but the most used are the mapping functions of Haldane (Haldane, 1919) and Kosambi (Kosambi, 1944). We choose the Kosambi mapping function because it assumes interference between crossover events. </div>

</p>

```{r kosambi, echo=TRUE, message=FALSE, warning=FALSE}
set_map_fun(type = "kosambi")
```

</p>

#### 3.3.2 Ordering of Markers:

<div align="justify">  The ordering process of the markers is considered an optimization problem (Traveling Salesman Problem - TSP) which seeks to find the best solution among all possible solutions. For $m$ markers, there are $m!/2$ possible orders to be investigated and in TSP it is impossible to obtain an exact solution when the number of markers is large. Thus, several methods for ordering molecular markers have proposed to solve the ordering problem such as Rapid Chain Delineation (RCD) (Doerge, 1996), Recombination Counting and Ordering (RECORD) (Van Os *et al*., 2005), Unidirectional Growth (UG) (Tan and Fu, 2006), and  multidimensional scaling (MDS) (Preedy and Hackett, 2016). Another criterion that can be used is the hidden Markov model (HMM) (Lander and Green, 1987). Through HMM, it is possible to obtain the maximum likelihood estimators (MLE) of the values using information from all the markers simultaneously (multipoint). </div>

</p>

<div align="justify"> The OneMap software has two-point approaches and multipoint approaches for estimating the order of markers. As the multipoint approach allows for the retrieval of information from the entire linkage map, we decided to use this approach. Therefore, the ordering of the markers within the linking groups was done using the **order_seq** function. </div>

</p>


```{r ord1, message=FALSE, warning=FALSE, results = "hide"}
LGs_hmm_safe = LGs_hmm_all = vector("list", length = c(lgs$n.groups))
    
    for (i in 1:lgs$n.groups) {
      
        LGs = make_seq(lgs, i)
        LG_hmm = order_seq(LGs, n.init = 5, subset.search = "twopt", twopt.alg = "rcd", THRES = 3)
        LG_safe = make_seq(LG_hmm, "safe")
        LG_all = make_seq(LG_hmm, "force")
        
        
        LGs_hmm_safe[[i]] = LG_safe
        LGs_hmm_all[[i]] = LG_all
        
    }
```

</p>

#### 3.3.3 Plotting the Recombination Fraction Matrix:

<div align="justify"> The ordering methods provide approximate solutions and may contain errors. Thus, additional tools can be useful in the final diagnosis of the marker ordering in a linkage group. </div>

</p>

<div align="justify"><div align="justify"> In OneMap, the ordering of markers within linkage groups can be diagnosed using heat maps. They are made by plotting the recombination matrix based on a color scale, in which, cool colors (blue areas) represent a weak linkage and hot colors (red areas) represent a strong linkage. Hot colors are expected to be concentrated along the diagonal while the cold colors fall far from it.  To get the first diagnosis of ordering, we plot a heat map for each linkage group using the rf_graph_table function.</div>

</p>

```{r graph1, echo=TRUE, fig.align='center', fig.height=3, fig.width=9, warning=FALSE}
g1 = rf_graph_table(LGs_hmm_all[[1]], mrk.axis = "none")
g2 = rf_graph_table(LGs_hmm_all[[2]], mrk.axis = "none")
grid.arrange(g1, g2, ncol=2)
```

```{r graph2, echo=TRUE, fig.align='center', fig.height=3, fig.width=9, warning=FALSE}
g3 = rf_graph_table(LGs_hmm_all[[3]], mrk.axis = "none")
g4 = rf_graph_table(LGs_hmm_all[[4]], mrk.axis = "none")
grid.arrange(g3, g4, ncol=2)
```

```{r graph3, echo=TRUE, fig.align='center', fig.height=3, fig.width=9, warning=FALSE}
g5 = rf_graph_table(LGs_hmm_all[[5]], mrk.axis = "none")
g6 = rf_graph_table(LGs_hmm_all[[6]], mrk.axis = "none")
grid.arrange(g5, g6, ncol=2)
```

```{r graph4, echo=TRUE, fig.align='center', fig.height=3, fig.width=9, warning=FALSE}
g7 = rf_graph_table(LGs_hmm_all[[7]], mrk.axis = "none")
g8 = rf_graph_table(LGs_hmm_all[[8]], mrk.axis = "none")
grid.arrange(g7, g8, ncol=2)
```

```{r graph5, echo=TRUE, fig.align='center', fig.height=3, fig.width=9, warning=FALSE}
g9 = rf_graph_table(LGs_hmm_all[[9]], mrk.axis = "none")
g10 = rf_graph_table(LGs_hmm_all[[10]], mrk.axis = "none")
grid.arrange(g9, g10, ncol=2)
```

```{r graph6, echo=TRUE, fig.align='center', fig.height=3, fig.width=9, warning=FALSE}
g11 = rf_graph_table(LGs_hmm_all[[11]], mrk.axis = "none")
g12 = rf_graph_table(LGs_hmm_all[[12]], mrk.axis = "none")
grid.arrange(g11, g12, ncol=2)
```

```{r graph7, echo=TRUE, fig.align='center', fig.height=3, fig.width=9, warning=FALSE}
g13 = rf_graph_table(LGs_hmm_all[[13]], mrk.axis = "none")
g14 = rf_graph_table(LGs_hmm_all[[14]], mrk.axis = "none")
grid.arrange(g13, g14, ncol=2)
```

<div align="justify"> Among all the linkage groups, group 3 and group 14 were the groups that most called our attention. Group 3 has an unusual pattern and is likely made up of several other groups. Group 14 has only 3 markers and these markers probably come from other groups. </div>
</p>

#### 3.3.4 Comparing the Maps:

<div align="justify"> To verify these hypotheses, we initially compared all linkage groups with all linkage groups from the published map. When we runned the analysis, we observed the following aspects: </div>

</p>

* <div align="justify">  nine groups correspond well with the published map (LG2, LG4, LG5, LG6, LG7, LG8, LG10, LG11, and LG12); </div>

</p>

* <div align="justify">  many markers of group 3 are present in groups 3, 7, and 10 of the published map;  </div>

</p>

* <div align="justify"> some markers from group 1 and 13 are present only in group 1 of the published map and some markers from group 9 and 14 are present only in group 14 of the published map; </div>

</p>

* <div align="justify"> one marker of group 11 is also present in group 2 of the published map. </div>


</p>

```{r map1, echo=TRUE, message=FALSE, warning=FALSE}
map_1 = list () 
    map_1[[1]] = c("CC138","BA153","AA173C","BA214","AAT267","BA129",
                  "CA389C","CA174","BA416","BA374","AAT225","AAT333")
    
    map_1[[2]] = c("CC385C","BA125","AA420","CC130","CA228","BC167",
                     "BA172","CB280","BA175","CC132","BD99","CA279")
    
    map_1[[3]] = c("CC378","BD286","CA220","CA238","CB156","BA75",
                  "BA384","CA131","AA404","BD429","BC586C","BC70")
    
    map_1[[4]] = c("CB329","CA399","CA415","CA233","AA374C","CA497",
                 "CA183C","AA166C","CA384","BD239","BA279C","BA113",
                           "BC321","CC270","CB257","AAT367","AA384")
    
    map_1[[5]] = c( "BB186","CC124","BA69","BB210","BD316C","AA378",
                  "AA268","CC342","AA454C","AA163","BC506","AA118C")
    
    map_1[[6]] = c("AAT230","BD179","AA277","AA158","BD169","CC171C",
                   "AAT300","CC381","CA283","CC126","BC392","AA371C")
    
    map_1[[7]] = c("BD209","CA258C","AA280","CB173","BB103C","AA361",
                                            "BD170","AAT261","BB102")
    
    map_1[[8]] = c("BA372","CA210","CYCB","CA305","CC450","AAT39",
               "AAT211","CB272","BA314","CA122","CC338C","AAT217",
                                          "CC359","BB167","BA158")
    
    map_1[[9]] = c("CA217","CA167","BB218","BC108","BA396C","LFY",
                  "BA311","AA95","CA196","CA261","CB115","AAT222",
                                                           "BC83")
    
    map_1[[10]] = c("CB309","AAT283","BD433","AA153C","AA100","BB259",
                      "BB122","AAT374","CA289","BB281","CA96","BA497",
                                             "CC392","AAT372","BB190")
    
    map_1[[11]] = c("CC61","BC526C","BA387","CC53","BD340","AAT356",
                             "BD100","BA196","BB124","CYCA","BC374")
    
    
    map_1[[12]] = c("BA394","CC114","BD143","BA245C","BA175","AAT308",
                    "AAT364","BC330","AA296C","CA378","CB216","CA140",
                    "BB182")
    
    map_1[[13]] = c( "BB198","BC199","BA145","AAT240","CB126","CB263C",
                                                "CB55","BA334","CB172")
    
    map_1[[14]] = c("BC542","CA267","BB216","BD251","CA75","BD371",
                                     "CC93","AA137","BC80","CC320")
```

```{r map2, echo=TRUE, message=FALSE, warning=FALSE}
map_2 = list () 
    map_2[[1]] = data_seg$Marker[LGs_hmm_all[[1]]$seq.num]
    map_2[[2]] = data_seg$Marker[LGs_hmm_all[[2]]$seq.num]
    map_2[[3]] = data_seg$Marker[LGs_hmm_all[[3]]$seq.num]
    map_2[[4]] = data_seg$Marker[LGs_hmm_all[[4]]$seq.num]
    map_2[[5]] = data_seg$Marker[LGs_hmm_all[[5]]$seq.num]
    map_2[[6]] = data_seg$Marker[LGs_hmm_all[[6]]$seq.num]
    map_2[[7]] = data_seg$Marker[LGs_hmm_all[[7]]$seq.num]
    map_2[[8]] = data_seg$Marker[LGs_hmm_all[[8]]$seq.num]
    map_2[[9]] = data_seg$Marker[LGs_hmm_all[[9]]$seq.num]
    map_2[[10]] = data_seg$Marker[LGs_hmm_all[[10]]$seq.num]
    map_2[[11]] = data_seg$Marker[LGs_hmm_all[[11]]$seq.num]
    map_2[[12]] = data_seg$Marker[LGs_hmm_all[[12]]$seq.num]
    map_2[[13]] = data_seg$Marker[LGs_hmm_all[[13]]$seq.num]
    map_2[[14]] = data_seg$Marker[LGs_hmm_all[[14]]$seq.num]
```

```{r comp1, echo=TRUE, message=FALSE, warning=FALSE}
compare = matrix(nrow = 14, ncol =14)
    
      for (i in 1:14) {
      
      map_2[[i]] 
      
          for (j  in 1:14)
            
          compare[i,j] = sum(map_2[[i]] %in% map_1[[j]])
      
      }
compare
```


#### 3.3.5 Checking the Linkage Groups:

</p>

##### Linkage Group 3:

<div align="justify"> These comparisons were very useful because from them we confirm our hypotheses, discover other information and make some decisions. The first decision was taken in relation to group 3. Initially, we removed some badly positioned markers using **drop_marker** function and we grouped these markers again. As expected, three groups were formed. Later, the markers of each group were ordered again.  </div>

</p>


```{r graph8, echo=TRUE, fig.align='center'}
rf_graph_table(LGs_hmm_all[[3]], mrk.axis = "numbers")
```

```{r LG3.1, echo=TRUE, message=FALSE, warning=FALSE}
LG3 = LGs_hmm_all[[3]]
LG3 = drop_marker(LG3, c(234,220,130, 236,403))
LG3 = map(LG3)
LG3 = group(LG3, LOD = LOD_sug)
```

```{r LG3.2, echo=TRUE, message=FALSE, warning=FALSE, results = "hide"}
LG3_hmm_safe = LG3_hmm_all = vector("list", length = c(LG3$n.groups))

for (i in 1:LG3$n.groups) {
      
      LGs = make_seq(LG3, i)
      
      LG3_hmm = order_seq(input.seq = LGs, n.init = 5,
                              subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 3)
      
      LG3_safe = make_seq(LG3_hmm, "safe")
      LG3_all = make_seq(LG3_hmm, "force")
      
      LG3_hmm_safe[[i]] = LG3_safe
      LG3_hmm_all[[i]] = LG3_all
      
    }
```

```{r grap9, echo=TRUE, fig.align='center', fig.height=3, fig.width=11}
g1 = rf_graph_table(LG3_hmm_all[[1]], mrk.axis = "none")
g2 = rf_graph_table(LG3_hmm_all[[2]], mrk.axis = "none")
g3 = rf_graph_table(LG3_hmm_all[[3]], mrk.axis = "none")
grid.arrange(g1, g2, g3, ncol=3)
```


```{r LG3.3, echo=TRUE, message=FALSE, warning=FALSE}
# Linkage groups of the new map (16 groups)
map_3 = list () 
    map_3[[1]] = data_seg$Marker[LGs_hmm_all[[1]]$seq.num]
    map_3[[2]] = data_seg$Marker[LGs_hmm_all[[2]]$seq.num]
    map_3[[3]] = data_seg$Marker[LG3_hmm_all[[1]]$seq.num]
    map_3[[4]] = data_seg$Marker[LG3_hmm_all[[2]]$seq.num]
    map_3[[5]] = data_seg$Marker[LG3_hmm_all[[3]]$seq.num]
    map_3[[6]] = data_seg$Marker[LGs_hmm_all[[4]]$seq.num]
    map_3[[7]] = data_seg$Marker[LGs_hmm_all[[5]]$seq.num]
    map_3[[8]] = data_seg$Marker[LGs_hmm_all[[6]]$seq.num]
    map_3[[9]] = data_seg$Marker[LGs_hmm_all[[7]]$seq.num]
    map_3[[10]] = data_seg$Marker[LGs_hmm_all[[8]]$seq.num]
    map_3[[11]] = data_seg$Marker[LGs_hmm_all[[9]]$seq.num]
    map_3[[12]] = data_seg$Marker[LGs_hmm_all[[10]]$seq.num]
    map_3[[13]] = data_seg$Marker[LGs_hmm_all[[11]]$seq.num]
    map_3[[14]] = data_seg$Marker[LGs_hmm_all[[12]]$seq.num]
    map_3[[15]] = data_seg$Marker[LGs_hmm_all[[13]]$seq.num]
    map_3[[16]] = data_seg$Marker[LGs_hmm_all[[14]]$seq.num]
```

```{r comp2, echo=TRUE, message=FALSE, warning=FALSE}
compare = matrix(nrow = 16, ncol =14)
    
    for (i in 1:16) {
      
      map_3[[i]] 
      
      for (j  in 1:14)
        
        compare[i,j] = sum(map_3[[i]] %in% map_1[[j]])
      
    }
compare
```


</p>

##### Joining Linkage Group (1 and 13; 9 and 14):

 <div align="justify">The second decision was made in relation to groups 1 and 13 and groups 9 and 14. The junctions of groups 1 and 13 and 9 and 14 were made and the markers were ordered again in their respective groups. </div>


```{r lg1.13.1, echo=TRUE}
LG1 = LGs_hmm_all[[1]]$seq.num
LG13 = LGs_hmm_all[[13]]$seq.num

seq1.13 = c(LG1, LG13)
seq1.13 = make_seq(twopts, seq1.13)
group_LG1.13 = group(seq1.13, LOD = 4)
```

```{r lg1.13.2, echo=TRUE, results='hide'}
LG1.13_order = order_seq(seq1.13)
LG1.13_order = make_seq(LG1.13_order, "force")
```

```{r lg1.13.3, echo=TRUE, fig.align='center'}
rf_graph_table(LG1.13_order, mrk.axis = "numbers")
```


```{r lg9.14.1, echo=TRUE}
LG9 = LGs_hmm_all[[9]]$seq.num
LG14 = LGs_hmm_all[[14]]$seq.num
    
seq9.14 = c(LG9, LG14)
seq9.14 = make_seq(twopts, seq9.14)
group_LG9.14 = group(seq9.14, LOD = 4)
```

```{r lg9.14.2, echo=TRUE, fig.align='center', results='hide'}
LG9.14_order = order_seq(seq9.14)
LG9.14_order = make_seq(LG9.14_order, "force")
```

```{r lg9.14.3, echo=TRUE, fig.align='center'}
rf_graph_table(LG9.14_order, mrk.axis = "numbers")
```

</p>

##### Linkage Group 11:

<div align="justify"> Finally, when analyzing the heatmap of group 11, we saw that marker 44 was misplaced and this is exactly the marker that corresponds to group 2 of the published map. </div>

```{r lg11.1, echo=TRUE, fig.align='center'}
rf_graph_table(LGs_hmm_all[[11]], mrk.axis = "numbers")
```

```{r lg11.2, echo=TRUE}
compare = map_2[[11]] %in% map_1[[2]]
position = which(compare == TRUE)
position
```

</p>

#### 3.3.6 Update of Linkage Groups:

<div align="justify"> Once the 14 groups were established, each group was carefully analyzed. The unreliable markers, along with any linked but unplaced markers, were then individually evaluated using the **drop_marker**, **map**, **try_seq**, **make_seq**, and **ripple_seq** functions. The purpose of the **ripple_seq** function is to evaluate alternative orders. </div>

</p>

```{r lgs, echo=TRUE}
LG1 = LG1.13_order
LG2 = LGs_hmm_all[[2]] 
LG3 = LG3_hmm_all[[2]]
LG4 = LGs_hmm_all[[4]]
LG5 = LGs_hmm_all[[5]]
LG6 = LGs_hmm_all[[6]]
LG7 = LG3_hmm_all[[3]]
LG8 = LGs_hmm_all[[7]]
LG9 = LGs_hmm_all[[8]]
LG10 = LG3_hmm_all[[1]]
LG11 = LGs_hmm_all[[10]]
LG12 = LGs_hmm_all[[11]]
LG13 = LGs_hmm_all[[12]]
LG14 = LG9.14_order
```

</p>

#### Linkage Group 1: 

```{r lg1.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG1, mrk.axis = "numbers")
```

```{r lg1.2, echo=TRUE, results='hide'}
LG1_drop1 = drop_marker(LG1, c(257, 1))
LG1_map1 = map(LG1_drop1)
ripple_seq(LG1_map1, ws = 5)
```

```{r lg1.3, echo=TRUE, results='hide'}
LG1_drop2 = drop_marker(LG1_map1, c(225,28,96))
LG1_map2 = map(LG1_drop2)
```

```{r lg1.4, echo=TRUE, results='hide', fig.align='center'}
LG1_extend_225 = try_seq(LG1_map2, 225)
LG1_extend_225 = make_seq(LG1_extend_225, 3, 1)
rf_graph_table(LG1_extend_225, mrk.axis = "numbers")
```

```{r lg1.5, echo=TRUE, results='hide', fig.align='center'}
LG1_extend_28 = try_seq(LG1_extend_225, 28)
LG1_extend_28 = make_seq(LG1_extend_28, 4, 1)
rf_graph_table(LG1_extend_28, mrk.axis = "numbers")
```

```{r lg1.6, echo=TRUE, results='hide', fig.align='center'}
LG1_extend_96 = try_seq(LG1_extend_28, 96)
LG1_extend_96 = make_seq(LG1_extend_96, 5, 1)
LG1_final = LG1_extend_96
rf_graph_table(LG1_final, mrk.axis = "numbers")
```


</p>

#### Linkage Group 2: 

```{r lg2.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG2, mrk.axis = "numbers")
```

```{r lg2.2, echo=TRUE, results='hide'}
LG2_drop1 = drop_marker(LG2, c(251,83))
LG2_map1 = map(LG2_drop1)
ripple_seq(LG2_map1, ws = 5)
LG2_final = LG2_map1
```

```{r lg2.3, echo=TRUE, fig.align='center'}
rf_graph_table(LG2_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 3: 

```{r lg3.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG3, mrk.axis = "numbers")
```

```{r lg3.2, echo=TRUE, fig.align='center', results='hide'}
LG3_drop1 = drop_marker(LG3, c(114,228,117,32,183))
LG3_map1 = map(LG3_drop1)
ripple_seq(LG3_map1, ws = 5)
```

```{r lg3.3, echo=TRUE, fig.align='center', results='hide'}
LG3_final = LG3_map1
rf_graph_table(LG3_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 4: 

```{r lg4.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG4, mrk.axis = "numbers")
```

```{r lg4.2, echo=TRUE, results='hide'}
LG4_drop1 = drop_marker(LG4, c(4))
LG4_map1 = map(LG4_drop1)
```


```{r lg4.3, echo=TRUE, results='hide'}
ripple_seq(LG4_map1, ws = 5)
LG4_drop2 = drop_marker(LG4_map1, c(238,110,291,125,76))
LG4_map2 = map(LG4_drop2)
```

```{r lg4.5, echo=TRUE, fig.align='center', results='hide'}
LG4_extend_238 = try_seq(LG4_map2, 238)
LG4_extend_238 = make_seq(LG4_extend_238, 1, 1)
rf_graph_table(LG4_extend_238, mrk.axis = "numbers")
```


```{r lg4.6, echo=TRUE, fig.align='center', results='hide'}
LG4_extend_110 = try_seq(LG4_extend_238, 110)
LG4_extend_110 = make_seq(LG4_extend_110, 2, 1)
rf_graph_table(LG4_extend_110, mrk.axis = "numbers")
```

```{r lg4.7, echo=TRUE, fig.align='center',  results='hide'}
LG4_extend_291 = try_seq(LG4_extend_110, 291)
LG4_extend_291 = make_seq(LG4_extend_291, 3, 1)
rf_graph_table(LG4_extend_291, mrk.axis = "numbers")
```

```{r lg4.8, echo=TRUE, fig.align='center', results='hide'}
LG4_final = drop_marker(LG4_extend_291, c(145,158))
LG4_final = map(LG4_final)
rf_graph_table(LG4_final, mrk.axis = "numbers")
```


</p>

#### Linkage Group 5: 

```{r lg5.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG5, mrk.axis = "numbers")
```

```{r lg5.2, echo=TRUE, results='hide'}
LG5_drop1 = drop_marker(LG5, c(191))
LG5_map1 = map(LG5_drop1)
ripple_seq(LG5_map1, ws = 5)
LG5_final = LG5_map1
```

```{r lg5.3, echo=TRUE, fig.align='center'}
rf_graph_table(LG5_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 6: 

```{r lg6.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG6, mrk.axis = "numbers")
```

```{r lg6.2, echo=TRUE, results='hide'}
LG6_drop1 = drop_marker(LG6, c(6,203,310,212,149))
LG6_map1 = map(LG6_drop1)
ripple_seq(LG6_map1, ws = 5)
LG6_final = LG6_map1
```

```{r lg6.3, echo=TRUE, fig.align='center'}
rf_graph_table(LG6_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 7: 

```{r lg7.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG7, mrk.axis = "numbers")
```

```{r lg7.2, echo=TRUE, results='hide', fig.align='center'}
LG7_drop1 = drop_marker(LG7, c(148,118,71,115))
LG7_map1 = map(LG7_drop1)
rf_graph_table(LG7_map1, mrk.axis = "numbers")
```

```{r lg7.3, echo=TRUE, results='hide'}
ripple_seq(LG7_map1, ws = 5)
LG7_final = LG7_map1
```

```{r lg7.4, echo=TRUE, fig.align='center'}
rf_graph_table(LG7_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 8: 

```{r lg8.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG8, mrk.axis = "numbers")
```

```{r lg8.2, echo=TRUE, results='hide'}
LG8_drop1 = drop_marker(LG8, c(46))
LG8_map1 = map(LG8_drop1)
ripple_seq(LG8_map1, ws = 5)
```

```{r lg8.3, echo=TRUE, fig.align='center'}
LG8_final = LG8_map1
rf_graph_table(LG8_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 9: 

```{r lg9.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG9, mrk.axis = "numbers")
```

```{r lg9.2, echo=TRUE, results='hide'}
LG9_drop1 = drop_marker(LG9, c(27,226,17))
LG9_map1 = map(LG9_drop1)
ripple_seq(LG9_drop1, ws = 5)
```

```{r lg9.3, echo=TRUE, fig.align='center', results='hide'}
LG9_extend_117 = try_seq(LG9_map1, 17)
LG9_extend_117 = make_seq(LG9_extend_117, 12, 1)
LG9_final = LG9_extend_117
rf_graph_table(LG9_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 10: 

```{r lg10.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG10, mrk.axis = "numbers")
```

```{r l10.2, echo=TRUE, results='hide'}
LG10_drop1 = drop_marker(LG10, c(255,107,406,327,123))
LG10_map1 = map(LG10_drop1)
ripple_seq(LG10_map1, ws = 5)
```

```{r lg10.3, echo=TRUE, fig.align='center'}
LG10_final = LG10_map1 
rf_graph_table(LG10_map1, mrk.axis = "numbers")
```

</p>

#### Linkage Group 11: 

```{r LG11.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG11, mrk.axis = "numbers")
```

```{r LG11.2, echo=TRUE, results='hide'}
ripple_seq(LG11, ws = 5)
```

</p>

#### Linkage Group 12: 

```{r lg12.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG12, mrk.axis = "numbers")
```

```{r l11.2, echo=TRUE, results='hide'}
LG12_drop1 = drop_marker(LG12, c(30,44, 193,204,241))
LG12_map1 = map(LG12_drop1)
ripple_seq(LG12_map1, ws = 5)
```

```{r lg11.3, echo=TRUE, fig.align='center'}
LG12_final = LG12_map1
rf_graph_table(LG12_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 13: 

```{r lg13.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG13, mrk.axis = "numbers")
```

```{r l13.2, echo=TRUE, results='hide'}
LG13_drop1 = drop_marker(LG13, c(116))
LG13_map1 = map(LG13_drop1)
ripple_seq(LG13_map1, ws = 5)
```

```{r lg13.3, echo=TRUE, fig.align='center'}
LG13_final = LG13_map1
rf_graph_table(LG13_final, mrk.axis = "numbers")
```

</p>

#### Linkage Group 14: 

```{r lg14.1, echo=TRUE, fig.align='center'}
rf_graph_table(LG14, mrk.axis = "numbers")
```


```{r l14.1, echo=TRUE, results='hide'}
ripple_seq(LG14, ws = 5)
```

```{r l14.2, echo=TRUE, results='hide'}
LG14_drop1 = drop_marker(LG14, c(215,99,258,21))
LG14_map1 = map(LG14_drop1)
```

```{r lg14.3, echo=TRUE, fig.align='center', results='hide'}
LG14_extend_21 = try_seq(LG14_map1, 21)
LG14_extend_21 = make_seq(LG14_extend_21, 2, 1)
rf_graph_table(LG14_extend_21, mrk.axis = "numbers")
```

```{r lg14.4, echo=TRUE, fig.align='center', results='hide'}
LG14_extend_215 = try_seq(LG14_extend_21, 215)
LG14_extend_215 = make_seq(LG14_extend_215, 1, 1)
rf_graph_table(LG14_extend_215, mrk.axis = "numbers")
```

```{r lg14.5, echo=TRUE, fig.align='center', results='hide'}
LG14_extend_99 = try_seq(LG14_extend_215, 99)
LG14_extend_99 = make_seq(LG14_extend_99, 1, 1)
rf_graph_table(LG14_extend_99, mrk.axis = "numbers")
```

```{r lg14.6, echo=TRUE, fig.align='center', results='hide'}
LG14_extend_258 = try_seq(LG14_extend_99, 258)
LG14_extend_258 = make_seq(LG14_extend_258, 1, 1)
LG14_final = LG14_extend_258
rf_graph_table(LG14_final, mrk.axis = "numbers")
```

</p>


### 7. Comparing the Final Map

```{r map4, echo=TRUE}
map_4 = list () 
map_4[[1]] = data_seg$Marker[LG1_final$seq.num]
map_4[[2]] = data_seg$Marker[LG2_final$seq.num]
map_4[[3]] = data_seg$Marker[LG3_final$seq.num]
map_4[[4]] = data_seg$Marker[LG4_final$seq.num]
map_4[[5]] = data_seg$Marker[LG5_final$seq.num]
map_4[[6]] = data_seg$Marker[LG6_final$seq.num]
map_4[[7]] = data_seg$Marker[LG7_final$seq.num]
map_4[[8]] = data_seg$Marker[LG8_final$seq.num]
map_4[[9]] = data_seg$Marker[LG9_final$seq.num]
map_4[[10]] = data_seg$Marker[LG10_final$seq.num]
map_4[[11]] = data_seg$Marker[LG11$seq.num]
map_4[[12]] = data_seg$Marker[LG12_final$seq.num]
map_4[[13]] = data_seg$Marker[LG13_final$seq.num]
map_4[[14]] = data_seg$Marker[LG14_final$seq.num]

```

```{r comp3, echo=TRUE}
compare = matrix(nrow = 14, ncol =14)

for (i in 1:14) {
  
  map_4[[i]] 
  
  for (j  in 1:14)
    
    compare[i,j] = sum(map_4[[i]] %in% map_1[[j]])
  
}

compare
```

<div align="justify"> After all manipulations were done, we compared the final map with the published map and observed that the two maps matched well. </div>

</p>

### 8. Drawing the Genetic Linkage Map

<div align="justify"> Once all linkage groups were obtained, the final map was drawn using the **draw_map2** function. The final map has 305 markers distributed across 14 linkage groups. Remembering that after the two filterings we were left with 356 markers, so 51 doubtful markers were removed from this analysis. </div>

</p>

```{r draw1, echo=TRUE, fig.align='center', results='hide'}
draw_map2(LG1_final, LG2_final, LG3_final, LG4_final, LG5_final, 
   LG6_final, LG7_final, LG8_final, LG9_final, LG10_final, LG11, 
                LG12_final, LG13_final, LG14_final, tag = "all", 
  group.names = c("LG1","LG2","LG3", "LG4", "LG5", "LG6", "LG7", 
              "LG8","LG9","LG10","LG11", "LG12","LG13", "LG14"),
                                      output = "map_final.pdf")
```


<img src="map1.png">

</p>


<img src="map2.png">

</p>

<img src="map3.png">

</p>

<img src="map4.png">

</p>


### References: 

</p>


<div align="justify"> Doerge, R. and B. Weir. Constructing genetic maps by rapid chain delineation. 1994. </div>

</p>

<div align="justify"> Fishman, L.A., Kelly J., Morgan, E., Willis J.H. A genetic map in the Mimulus guttatus species complex reveals transmission ratio distortion due to heterospecific interactions. **Genetics 159**:1701-1716, 2001.  </div>

</p>

<div align="justify"> Haldane, J. B. S. The combination of linkage values, and the calculation of distances between the loci of linked factors. **Hournal of Gnetics 8**: 299–309, 1919. </div>

</p>

<div align="justify"> Kosambi, D. D. The estimation of map distances from recombination values. volume 12, pp. 172–175, **Annals of Eugenics**, 1944. </div>

</p>

<div align="justify"> Lander, E. S. and Green, P. Construction of multilocus genetic linkage maps in humans. **Proc. Natl. Acad. Sci. USA 84**, 2363-2367, 1987. </div>

</p>

<div align="justify"> Margarido G. R. A., Souza A. P., Garcia A. A. F., 2007. OneMap: software for genetic mapping in outcrossing species. **Hereditas 144**: 78–79. </div>

</p>

<div align="justify"> Preedy, K. F. and C. A. Hackett. A rapid marker ordering approach for high-density genetic linkage maps in experimental autotetraploid populations using multidimensional scaling. **Theoretical and Applied Genetics 129**: 2117–2132, 2016. </div>

</p>

<div align="justify"> Tan, Y.-D. and Y.-X. Fu. A novel method for estimating linkage maps. **Genetics 173**: 2383–2390, 2006. </div>

</p>

<div align="justify"> Van Os, H., P. Stam, R. G. Visser, and H. J. Van Eck. Record: a novel method for ordering loci on a genetic linkage map. **Theoretical and Applied Genetics 112**: 30–40, 2005.

</div>

</p>

<div align="justify"> Wu, R., 2002 Simultaneous Maximum Likelihood Estimation of Linkage and Linkage Phases in Outcrossing Species. **Theoretical population biology 363**: 349–363. </div>





